petab/visualize/helper_functions.py

```diff
@@ -390,56 +396,208 @@ def get_default_vis_specs(
+    return exp_data, columns_dict
+
+
+def expand_vis_spec_settings(vis_spec, columns_dict):
+    """
+    only makes sense if DATASET_ID is not in vis_spec.columns?
+
+    Returns:
+        A visualization specification DataFrame
+    """
+    columns_to_expand = [PLOT_NAME, PLOT_TYPE_SIMULATION, PLOT_TYPE_DATA,
+                         X_VALUES, X_OFFSET, X_LABEL, X_SCALE, Y_OFFSET,
+                         Y_LABEL, Y_SCALE, LEGEND_ENTRY]
+
+    for column in vis_spec.columns:
+        if column in columns_to_expand:
+            column_entries = []
+            if Y_VALUES in vis_spec.columns:
+                for i, plot_id in enumerate(columns_dict[PLOT_ID]):
+                    select_conditions = (vis_spec[PLOT_ID] == plot_id) & (
+                        vis_spec[Y_VALUES] == columns_dict[Y_VALUES][i])
+                    column_entries.append(
+                        vis_spec[select_conditions].loc[:, column].values[0])
+            else:
+                for plot_id in columns_dict[PLOT_ID]:
+                    select_conditions = vis_spec[PLOT_ID] == plot_id
+                    column_entries.append(
+                        vis_spec[select_conditions].loc[:, column].values[0])
+            columns_dict[column] = column_entries
+    vis_spec = pd.DataFrame(columns_dict)
+    return vis_spec
+
+
+def create_or_update_vis_spec(
+        exp_data: pd.DataFrame,
+        exp_conditions: pd.DataFrame,
+        vis_spec: Optional[pd.DataFrame] = None,
+        dataset_id_list: Optional[List[IdsList]] = None,
+        sim_cond_id_list: Optional[List[IdsList]] = None,
+        sim_cond_num_list: Optional[List[NumList]] = None,
+        observable_id_list: Optional[List[IdsList]] = None,
+        observable_num_list: Optional[List[NumList]] = None,
+        plotted_noise: Optional[str] = MEAN_AND_SD):
+    """
+    Helper function for plotting data and simulations, which updates vis_spec
+    file if necessary or creates a default visualization table and
+    updates/creates DATASET_ID column of exp_data. As a result, a visualization
+    specification file exists with columns PLOT_ID, DATASET_ID, Y_VALUES and
+    LEGEND_ENTRY
+
+    Returns:
+        A tuple of visualization specification DataFrame and experimental
+        DataFrame.
+    """
+    if vis_spec is None:
+        # create dataframe
+        exp_data, columns_dict = \
+            get_vis_spec_dependent_columns_dict(exp_data,
+                                                exp_conditions,
+                                                dataset_id_list,
+                                                sim_cond_id_list,
+                                                sim_cond_num_list,
+                                                observable_id_list,
+                                                observable_num_list)
+        vis_spec = pd.DataFrame(columns_dict)
+    else:
+        # TODO: do validation
+        # so, plotid is definitely there
```

Comment:
validation meaning something like `PLOT_ID in vis_spec.columns`?

Reply:
Yes and #190

Reply:
ah I see. you could link this issue in the comment